src/core/lib/event_engine/thread_pool/work_stealing_thread_pool.cc

```diff
@@ -268,15 +268,15 @@ void WorkStealingThreadPool::WorkStealingThreadPoolImpl::Lifeguard::
     MaybeStartNewThread();
   }
   lifeguard_running_ = false;
-  lifeguard_shutdown_cv_.Signal();
+  lifeguard_is_shut_down_cv_.Signal();
 }
 
 void WorkStealingThreadPool::WorkStealingThreadPoolImpl::Lifeguard::
     BlockUntilShutdown() {
   grpc_core::MutexLock lock(&lifeguard_shutdown_mu_);
   while (lifeguard_running_) {
-    lifeguard_shutdown_cv_.Signal();
-    lifeguard_shutdown_cv_.WaitWithTimeout(
+    lifeguard_should_shut_down_cv_.Signal();
```

Comment:
We should be able to move this signal to be just when the shutdown bit changes, right?

(doing so would help catch bugs sooner by not spinning in a wakeup loop)

Reply:
If we only signal once, I think it's possible for the lifeguard main loop to miss it and wait longer than necessary (whatever the current backoff time is). This repeated signalling was a benchmark-driven optimization as well.

Reply:
I think this bit needs to be fixed prior to de-experimentalizing: this wakeup ought to be possible with a single signal when the flag changes.

That we need a loop here will likely lead to some wildly varying and load-dependent shutdown times.

(it may be that we need to accept a small perf regression here to achieve that, and I think the robustness improvement will be worth it)

Reply:
(happy for this to be a followup PR however)

Reply:
Replaced the CV+Mutex with Notifications, which permits a single signal.